### DISCUSSION TITLE
Best practice showing users local date/time in inertia vue

### DISCUSSION THREAD

**Discussion by j-codes-4** [inertiajs/inertia#424] [Help (Vue)] [answered]
Just wondering what everyone is doing to format date/time to users local timzone and format?  In a standard laravel app, we have casts and mutators we can use based on users settings.  On the vue side we have moment or vue-moment as some good options.  I will basically have time based tasks/projects/etc. that will need to be local to the user.  I'm not sure which option is better/easier so if anybody has an opinion or insight on this, please share.  Thanks for reading

**Comment by grantholle** [comment]
You will need to pass all dates and times to your pages in ISO 8601 format, then convert them to moment objects.

```php
$model->created_at->format('c');
```

```js
const createdAt = moment(model.created_at)
```

This gets much more tricky as your users are across different timezones. If you feel more comfortable on the Laravel side, then I would pass all dates and times in the format you already want them to be displayed on the frontend.

**Comment by claudiodekker** [comment]
Personally, what I tend to do most of the time, is just use Carbon on the back-end for formatting date/time strings, after which I use them in my front-end directly without mutating them. Something like this:
```php
return Inertia::render('Example', [
    'createdAgo' => $post->created_at->diffForHumans()
]);
```

However, when I need it to be updated on the front-end reactively depending on (for example) form changes (e.g. extracting the weekday from a selected date string), I do use something like `moment` ([or](https://github.com/you-dont-need/You-Dont-Need-Momentjs) even better: [day.js](https://github.com/iamkun/dayjs)) on the front-end, while using [`Carbon`](https://carbon.nesbot.com/docs/#api-formatting)'s `timestamp` method to pass the date/time to the front-end as an unix timestamp (similar to what @grantholle suggested), in an effort to try and avoid timezone mumbo-jumbo altogether and always just use UTC / server-time:

```php
return Inertia::render('Example', [
    'createdAt' => $post->created_at->timestamp
]);
```

```js
export default {
  props: {
    createdAt: String
  },
  computed: {
    weekday() {
      return dayjs(this.createdAt).day()
    }
  }
}
```

If you want to avoid using any third-part JS libraries altogether, you could also use something that I saw @reinink do the other day, which is use the natively supported `Intl.DateTimeFormat` object instead:
```js
Intl.DateTimeFormat('en', { dateStyle: 'full' }).formatToParts(Date.parse('2020-01-12 15:22:33'))

// Or, when working with timestamps directly:
Intl.DateTimeFormat('en', { dateStyle: 'full' }).formatToParts(this.createdAt)
Intl.DateTimeFormat('en', { dateStyle: 'full' }).formatToParts(1578838953000)
```

Which would give you something like this:
![Screenshot 2021-01-12 at 15 53 06](https://user-images.githubusercontent.com/1752195/104330454-461a2400-54ee-11eb-8f60-bc0ddfb1b0cc.png)

**Comment by j-codes-4** [comment, marked as answer]
Great answers, thank you both for taking the time to respond.  I will probably end up handling it mostly on the server side using carbon and mutators, but day.js looks nice also so I may end up using that in some situations.
